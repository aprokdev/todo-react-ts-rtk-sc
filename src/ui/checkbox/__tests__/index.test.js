import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { act, create } from 'react-test-renderer';
import Checkbox from '../index';

const props = {
    checked: false,
    className: 'test',
    id: 'test',
    onChange: () => null,
};

describe('Checkbox', () => {
    test('matches snapshot', () => {
        let tree;
        act(() => {
            tree = create(<Checkbox {...props} />);
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('add passed className prop to the end of classlist', () => {
        const className = 'test';
        render(<Checkbox {...props} className={className} />);
        const component = screen.getByTestId('checkbox');
        const classListLength = component.classList.length;
        const lastClassName = component.classList[classListLength - 1];
        expect(lastClassName).toBe(className);
    });

    test('onKeyDown event on focused <label> with Enter key fires onChange event in input', async () => {
        const mockChange = jest.fn();
        const user = userEvent.setup();
        render(<Checkbox onChange={mockChange} id="test" />);
        const square = screen.getByTestId('checkbox-square');
        const checkbox = screen.getByTestId('checkbox-input');
        square.focus();
        await user.keyboard('[Enter]');
        expect(mockChange).toHaveBeenCalledTimes(1);
        expect(checkbox.checked).toBe(true);
    });

    test('onClick event on <label> fires onChange event in input', async () => {
        const mockChange = jest.fn();
        const user = userEvent.setup();
        render(<Checkbox onChange={mockChange} id="test" />);
        const square = screen.getByTestId('checkbox-square');
        const checkbox = screen.getByTestId('checkbox-input');
        await user.click(square);
        expect(mockChange).toHaveBeenCalledTimes(1);
        expect(checkbox).toBeChecked();
    });
});
