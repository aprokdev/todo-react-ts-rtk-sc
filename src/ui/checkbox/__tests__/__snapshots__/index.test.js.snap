// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox matches snapshot 1`] = `
<span
  className="checkbox test"
  data-testid="checkbox"
>
  <input
    checked={false}
    className="checkbox__input"
    data-testid="checkbox-input"
    id="test"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="checkbox__square"
    data-testid="checkbox-square"
    htmlFor="test"
    onKeyDown={[Function]}
    tabIndex={0}
  />
</span>
`;
